File: backend/src/main/java/com/technicalescaperoom/backend/service/HintService.java
package com.technicalescaperoom.backend.service;

import com.technicalescaperoom.backend.config.security.PlayerPrincipal;
import com.technicalescaperoom.backend.dto.player.PlayerHintDto;
import com.technicalescaperoom.backend.dto.player.PlayerHintsResponseDto;
import com.technicalescaperoom.backend.dto.player.HintUseResponseDto;
import com.technicalescaperoom.backend.entity.Hint;
import com.technicalescaperoom.backend.entity.HintUsage;
import com.technicalescaperoom.backend.entity.Level;
import com.technicalescaperoom.backend.entity.Team;
import com.technicalescaperoom.backend.entity.TeamLevelProgress;
import com.technicalescaperoom.backend.entity.TeamStageProgress;
import com.technicalescaperoom.backend.enums.LevelStatus;
import com.technicalescaperoom.backend.exception.ResourceNotFoundException;
import com.technicalescaperoom.backend.repository.HintRepository;
import com.technicalescaperoom.backend.repository.LevelRepository;
import com.technicalescaperoom.backend.repository.TeamLevelProgressRepository;
import com.technicalescaperoom.backend.repository.TeamRepository;
import com.technicalescaperoom.backend.repository.HintUsageRepository;
import com.technicalescaperoom.backend.repository.TeamStageProgressRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Optional;
import java.util.stream.Collectors;

@Slf4j
@Service
@RequiredArgsConstructor
public class HintService {

    private final TeamRepository teamRepository;
    private final TeamLevelProgressRepository teamLevelProgressRepository;
    private final LevelRepository levelRepository;
    private final HintRepository hintRepository;
    private final HintUsageRepository hintUsageRepository;
    private final TeamStageProgressRepository teamStageProgressRepository;
    private final GameWebSocketPublisher webSocketPublisher;
    private final ScoringService scoringService;

    @Transactional(readOnly = true)
    public PlayerHintsResponseDto getHintsForPlayer(PlayerPrincipal principal) {
        if (principal == null) {
            throw new ResourceNotFoundException("No authenticated player session found.");
        }

        Team team = teamRepository.findById(principal.getTeamId())
                .orElseThrow(() -> new ResourceNotFoundException("Team not found for ID " + principal.getTeamId()));

        List<HintUsage> usedHints = hintUsageRepository.findByTeamId(team.getId());
        java.util.Set<String> usedHintKeys = usedHints.stream()
                .map(u -> u.getLevel().getLevelNumber() + "_" + u.getStageNumber())
                .collect(Collectors.toSet());

        List<Level> activeLevels = levelRepository.findByIsActiveTrueOrderByLevelNumberAsc();
        List<PlayerHintDto> hintDtos = new ArrayList<>();
        int unlockedCount = 0;

        for (Level level : activeLevels) {
            List<Hint> levelHints = hintRepository.findByLevelIdOrderByDisplayOrderAsc(level.getId());
            Map<Integer, Hint> stageHintMap = new HashMap<>();
            for (Hint h : levelHints) {
                if (Boolean.TRUE.equals(h.getIsActive())) {
                    stageHintMap.putIfAbsent(h.getStageNumber(), h);
                }
            }

            int stagesInLevel = switch (level.getLevelNumber()) {
                case 1, 2, 3 -> 2;
                case 4, 5, 6 -> 3;
                default -> 2;
            };

            for (int stageNum = 1; stageNum <= stagesInLevel; stageNum++) {
                Hint hint = stageHintMap.get(stageNum);
                String key = level.getLevelNumber() + "_" + stageNum;
                boolean isUnlocked = usedHintKeys.contains(key);
                String content = (isUnlocked && hint != null) ? hint.getHintContent() : null;

                if (isUnlocked) {
                    unlockedCount++;
                }

                PlayerHintDto dto = PlayerHintDto.builder()
                        .levelNumber(level.getLevelNumber())
                        .stageNumber(stageNum)
                        .hintNumber(1)
                        .isUnlocked(isUnlocked)
                        .hintContent(content)
                        .build();

                hintDtos.add(dto);
            }
        }

        return PlayerHintsResponseDto.builder()
                .hints(hintDtos)
                .unlockedCount(unlockedCount)
                .totalCount(hintDtos.size())
                .build();
    }

    @Transactional
    public HintUseResponseDto useHint(PlayerPrincipal principal, Integer levelNumber, Integer stageNumber, Integer hintNumber) {
        if (principal == null) throw new ResourceNotFoundException("No authenticated player session found.");
        Team team = teamRepository.findById(principal.getTeamId())
                .orElseThrow(() -> new ResourceNotFoundException("Team not found."));
        Level level = levelRepository.findByLevelNumber(levelNumber)
                .orElseThrow(() -> new ResourceNotFoundException("Level not found."));

        boolean alreadyUsed = hintUsageRepository.existsByTeamIdAndLevelIdAndStageNumberAndHintNumber(
                team.getId(), level.getId(), stageNumber, 1);

        List<Hint> hints = hintRepository.findByLevelIdAndStageNumberOrderByDisplayOrderAsc(level.getId(), stageNumber);
        if (hints.isEmpty()) {
            hints = hintRepository.findByLevelIdOrderByDisplayOrderAsc(level.getId());
        }
        int targetHintNumber = (hintNumber != null && hintNumber > 0) ? hintNumber : 1;
        Hint hint = hints.stream()
                .filter(item -> Boolean.TRUE.equals(item.getIsActive()))
                .filter(item -> item.getDisplayOrder() != null && item.getDisplayOrder().equals(targetHintNumber))
                .findFirst()
                .orElseThrow(() -> new ResourceNotFoundException("Hint " + targetHintNumber + " not configured for Level " + levelNumber + " Stage " + stageNumber));

        if (!alreadyUsed) {
            try {
                hintUsageRepository.saveAndFlush(HintUsage.builder()
                        .team(team)
                        .level(level)
                        .stageNumber(stageNumber)
                        .hintNumber(1)
                        .build());
                scoringService.recordHintUsage(team.getId(), principal.getPlayerId(), levelNumber, stageNumber, 1);
            } catch (org.springframework.dao.DataIntegrityViolationException e) {
                log.info("Concurrent hint usage detected for team {} level {} stage {}: safely treating as already used",
                        team.getId(), level.getId(), stageNumber);
                alreadyUsed = true;
            }
            webSocketPublisher.notifyHintUnlocked(team.getId(), levelNumber, 1);
        }

        return HintUseResponseDto.builder()
                .levelNumber(levelNumber)
                .stageNumber(stageNumber)
                .hintNumber(1)
                .hintContent(hint.getHintContent())
                .alreadyUsed(alreadyUsed)
                .build();
    }
}

File: backend/services/hintService.js
const teamRepository = require('../repositories/teamRepository');
const levelRepository = require('../repositories/levelRepository');
const hintRepository = require('../repositories/hintRepository');
const hintUsageRepository = require('../repositories/hintUsageRepository');
const teamLevelProgressRepository = require('../repositories/teamLevelProgressRepository');
const teamStageProgressRepository = require('../repositories/teamStageProgressRepository');
const scoringService = require('./scoringService');
const webSocketService = require('./webSocketService');
const {
  ResourceNotFoundException,
  EventUnavailableException,
  InvalidLevelTransitionException
} = require('../middleware/errorHandler');

class HintService {
  async getHintsForPlayer(principal) {
    if (!principal) throw new ResourceNotFoundException('No authenticated player session found.');
    const team = await teamRepository.findById(principal.teamId);
    if (!team) throw new ResourceNotFoundException('Team not found.');

    const usedHints = await hintUsageRepository.findByTeamId(team.id);
    const usedHintKeys = new Set(usedHints.map(u => `${u.levelNumber}_${u.stageNumber}`));

    const activeLevels = await levelRepository.findAllActive();
    const hintDtos = [];
    let unlockedCount = 0;

    for (const level of activeLevels) {
      const levelHints = await hintRepository.findByLevelIdOrderByDisplayOrderAsc(level.id);
      const stageHintMap = new Map();
      for (const h of levelHints) {
        if (!stageHintMap.has(h.stageNumber)) {
          stageHintMap.set(h.stageNumber, h);
        }
      }

      const stagesInLevel = (level.levelNumber >= 4) ? 3 : 2;
      for (let s = 1; s <= stagesInLevel; s++) {
        const hint = stageHintMap.get(s);
        const key = `${level.levelNumber}_${s}`;
        const isUnlocked = usedHintKeys.has(key);
        const content = (isUnlocked && hint) ? hint.hintContent : null;

        if (isUnlocked) unlockedCount++;

        hintDtos.push({
          levelNumber: level.levelNumber,
          stageNumber: s,
          hintNumber: 1,
          isUnlocked,
          hintContent: content
        });
      }
    }

    return {
      hints: hintDtos,
      unlockedCount,
      totalCount: hintDtos.length
    };
  }

  async useHint(principal, levelNumber, stageNumber, hintNumber = 1) {
    if (!principal) throw new ResourceNotFoundException('No authenticated player session found.');
    const team = await teamRepository.findById(principal.teamId);
    if (!team) throw new ResourceNotFoundException('Team not found.');

    if (team.gameState === 'COMPLETED') {
      throw new EventUnavailableException('Game is not currently active for hint requests.');
    }

    const level = await levelRepository.findByLevelNumber(levelNumber);
    if (!level) throw new ResourceNotFoundException(`Level ${levelNumber} not found.`);

    // Server-authoritative progression boundary check: prevent requesting hints for future levels or stages
    const progressList = await teamLevelProgressRepository.findByTeamIdOrderByLevelIdAsc(team.id);
    const activeProgress = progressList.find(
      p => p.levelStatus === 'AVAILABLE' || p.levelStatus === 'IN_PROGRESS'
    );

    let activeLevelNumber = 1;
    let activeStageNumber = 1;
    if (activeProgress) {
      activeLevelNumber = activeProgress.level.levelNumber;
      const questionAnswerService = require('./questionAnswerService');
      activeStageNumber = await questionAnswerService.findCurrentStage(team.id, activeProgress.levelId, activeLevelNumber);
    }

    if (levelNumber > activeLevelNumber || (levelNumber === activeLevelNumber && stageNumber > activeStageNumber)) {
      throw new InvalidLevelTransitionException('Cannot request hints for future levels or stages before reaching them.');
    }

    const alreadyUsed = await hintUsageRepository.existsByTeamIdAndLevelIdAndStageNumberAndHintNumber(
      team.id, level.id, stageNumber, 1
    );

    const hints = await hintRepository.findByLevelIdAndStageNumberOrderByDisplayOrderAsc(level.id, stageNumber);
    const targetHintNumber = (hintNumber && hintNumber > 0) ? hintNumber : 1;
    const hint = hints.find(h => h.displayOrder === targetHintNumber) || hints[0];

    if (!hint) {
      throw new ResourceNotFoundException(`Hint ${targetHintNumber} not configured for Level ${levelNumber} Stage ${stageNumber}`);
    }

    if (!alreadyUsed) {
      await hintUsageRepository.recordUsage(team.id, level.id, stageNumber, 1);
      await scoringService.recordHintUsage(team.id, principal.playerId, levelNumber, stageNumber, 1);
      webSocketService.notifyHintUnlocked(team.id, levelNumber, 1);
    }

    return {
      levelNumber,
      stageNumber,
      hintNumber: 1,
      hintContent: hint.hintContent,
      alreadyUsed
    };
  }
}

module.exports = new HintService();
